evm/src/arithmetic/shift.rs

```diff
@@ -0,0 +1,394 @@
+//! Support for the EVM SHL and SHR instructions.
+//!
+//! This crate verifies an EVM shift instruction, which takes two
+//! 256-bit inputs S and A, and produces a 256-bit output C satisfying
+//!
+//!    C = A << S (mod 2^256) for SHL or
+//!    C = A >> S (mod 2^256) for SHR.
+//!
+//! The way this computation is carried is by providing a third input
+//!    B = 1 << S (mod 2^256)
+//! and then computing:
+//!    C = A * B (mod 2^256) for SHL or
+//!    C = A / B (mod 2^256) for SHR
+//!
+//! Inputs A, S, and B, and output C, are given as arrays of 16-bit
+//! limbs. For example, if the limbs of A are a[0]...a[15], then
+//!
+//!    A = \sum_{i=0}^15 a[i] β^i,
+//!
+//! where β = 2^16 = 2^LIMB_BITS. To verify that A, S, B and C satisfy
+//! the equations, we proceed similarly to MUL for SHL and to DIV for SHR.
+
+use ethereum_types::U256;
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::{Field, PrimeField64};
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+
+use super::modular::modular_constr_poly_ext_circuit;
+use crate::arithmetic::columns::{self, *};
+use crate::arithmetic::modular::{generate_modular_op, modular_constr_poly};
+use crate::arithmetic::utils::*;
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+/// Generates a shift operation (either SHL or SHR).
+/// The inputs are stored in the form `(shift, input, 1 << shift)`.
+/// NB: if `shift > 2^32`, then the third register holds 0.
+pub fn generate<F: PrimeField64>(
+    lv: &mut [F],
+    nv: &mut [F],
+    is_shl: bool,
+    input: U256,
+    shift: U256,
+    result: U256,
+) {
+    // TODO: It would probably be clearer/cleaner to read the U256
+    // into an [i64;N] and then copy that to the lv table.
+    // The first input is the shift we need to apply.
+    u256_to_array(&mut lv[INPUT_REGISTER_0], shift);
+    // The second register holds the input which needs shifting.
+    u256_to_array(&mut lv[INPUT_REGISTER_1], input);
+    u256_to_array(&mut lv[OUTPUT_REGISTER], result);
+    // If `shift > 2^32`, the shifted displacement is set to 0.
+    // Compute 1 << shift and store it in the third input register.
+    let shifted_displacement = if shift > U256::from(255u64) {
+        U256::zero()
+    } else {
+        U256::one() << shift
+    };
+
+    u256_to_array(&mut lv[INPUT_REGISTER_2], shifted_displacement);
+
+    let input0 = read_value_i64_limbs(lv, INPUT_REGISTER_1); // input
+    let input1 = read_value_i64_limbs(lv, INPUT_REGISTER_2); // 1 << shift
+
+    if is_shl {
```

Comment:
I think the code in the two parts of this if statement can easily be refactored with the code in `mul.rs` and `divmod.rs` that they were copied from.